### ISSUE TITLE
Editor type safety: type safety requires compile

### ISSUE THREAD

**Issue by justinwaite** [tommedema/serverless-mono-example#4]
Another quick issue: you mentioned in the README that this had editor type safety that did not require a compile every time there is a change. I'm using VS Code and in order for it to have any type checking, it requires me to compile the projects first. Any thoughts on this? Are you using any special extensions?

Edit: for further information, Webstorm behaves in the same way -- a build is required before typechecking can occur.

Edit 2: I updated to Webstorm 2018.2 and it appears to do type checking without needing to recompile -- vs code is still an issue after updating.

**Comment by tommedema** [comment]
@jdeanwaite you're right, actually. I just confirmed this issue myself.

Turns out my code was recently compiled and I had mistakenly assumed that vscode correctly picked up the "runtime" type checks.

Not sure why vscode doesn't pickup the project. Perhaps we should open up an issue over at vscode.

**Comment by tommedema** [comment]
Reported at https://github.com/Microsoft/vscode/issues/57242

**Comment by jdeanwaite** [comment]
Thanks for looking into this! If you are a Webstorm user, 2018.2 works great with typechecking without compiling (they had some release notes specific to Typescript 3 features they implemented).

**Comment by tommedema** [comment]
I gave Webstorm a try but am seeing all sorts of issues there:

<img width="1489" alt="screen shot 2018-08-27 at 12 52 58 pm" src="https://user-images.githubusercontent.com/331833/44670146-46d2be00-a9f8-11e8-87ed-b3050398efe1.png">

I would be interested in getting this example repo to work well with Webstorm. However, adding `es2015` to lib should not be necessary when `es2017` is already there? Even after adding it the errors did not go away.

**Comment by jdeanwaite** [comment]
Make sure your Node.JS assistance is on and the JavaScript language version is set to ECMAScript 6:
![image](https://user-images.githubusercontent.com/8517615/44671932-c1411500-a9db-11e8-9582-55fde2523d4d.png)
![image](https://user-images.githubusercontent.com/8517615/44671935-c3a36f00-a9db-11e8-97f4-beafb26eaec4.png)

**Comment by tommedema** [comment]
@jdeanwaite I see. Are you sure the run-time typechecking works across references?

I just updated `random` to return a string:

<img width="549" alt="screen shot 2018-08-27 at 2 25 27 pm" src="https://user-images.githubusercontent.com/331833/44674877-3ffe7800-aa05-11e8-87c3-263619c523ac.png">

But did not see any errors in the consuming `sls-random`:

<img width="1438" alt="screen shot 2018-08-27 at 2 26 16 pm" src="https://user-images.githubusercontent.com/331833/44674889-4856b300-aa05-11e8-9a81-06400ce71ab0.png">

it's supposed to complain that `fetchNumber` returns a string and not a number

**Comment by tommedema** [comment]
I can now reproduce this issue. Not sure when it started appearing..

Not sure about a fix yet either.

**Comment by jdeanwaite** [comment]
@tommedema 

First off, in order to see the useful information, hold down CMD (if you're on a mac) and hover over the function to see details:

![image](https://user-images.githubusercontent.com/8517615/44679133-f3a83d80-a9ee-11e8-9a40-4ff040e3f0fd.png)

And you are right, it was only partially working. For instance, if I change a parameter, the new parameter shows up right away in the hints (type CMD + P on a mac while the cursor is inside of the parentheses to get hints). You can see in this image I added `testing: number`

![image](https://user-images.githubusercontent.com/8517615/44679216-30743480-a9ef-11e8-8940-404fb1f19dcc.png)

However, return types were not correctly updating (like what you showed).

**Comment by jdeanwaite** [comment]
One thing you CAN do is do your work while running `tsc -b -w` to watch the files while you code. This shows up in a reasonable amount of time (a few seconds):
![image](https://user-images.githubusercontent.com/8517615/44679364-a1b3e780-a9ef-11e8-9ffd-c71efd69832b.png)

I actually added `serverless-offline` to the package and the `serverless.yml` script to be able to test locally easier. I paired this with `tsc -b -w` using `concurrently`:

package.json:
```
  ...
  "scripts": {
    "offline": "concurrently --names \"WATCH, OFFLINE\" -c \"green,cyan\" \"yarn watch\" \"serverless offline start\""
  },
  ...
  "devDependencies": {
    ...
    "concurrently": "4.0.0",
    ...
  }
  ...
```

serverless.yml
```
plugins:
  - serverless-plugin-scripts
  - serverless-offline
```

**Comment by tommedema** [comment]
@jdeanwaite I tried `tsc -b -w` but it was getting stuck as soon as there was a compilation error somewhere. Editors on the other hand will do a best-effort and continue parsing the "broken" file for further typing information.

When you're doing a large refactor, you know that most projects won't compile anymore; but to make them compile you'd want to have typing information from other packages. `tsc -b -w` does not really help here.

Do you get my meaning?

**Comment by jdeanwaite** [comment]
It's true in the circumstances of refactoring it is not very helpful -- however I haven't had any issues with it getting "stuck" when a file failed to compile. After I fix the issue it seems to move on just fine.

**Comment by tommedema** [comment]
@jdeanwaite let's wait for https://github.com/Microsoft/vscode/issues/57242 to get some attention

**Comment by jdeanwaite** [comment]
Yes, sounds like a plan. Also I saw you posted an inquiry on https://github.com/prisma/serverless-plugin-typescript/issues/110.. This would be better than using `tsc -w`.
